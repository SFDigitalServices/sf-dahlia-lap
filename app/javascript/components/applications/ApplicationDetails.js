import React from 'react'
import ApplicationDetailsContentCard from './ApplicationDetailsContentCard'
import ApplicationDetailsContentTable from './ApplicationDetailsContentTable'

const ApplicationDetails = ({ application, fields, file_base_url }) => {

  const editLink = () => {
    if (!application.Is_Lottery_Complete && application.Application_Submission_Type == 'Paper') {
      return (
        <p><a href={`/applications/${application.Id}/edit`}>Edit Application</a></p>
      )
    }
  }

  let proofList = _.map(application.proof_files, (file) => {
    return <li key={file.Id}><a target='_blank' href={`${file_base_url}/servlet/servlet.FileDownload?file=${file.Id}`}>{file.Name}</a></li>
  })

  return (
    <div>
      {editLink()}
      <ApplicationDetailsContentCard
        dataCollection={application}
        title='Application Data'
        fields={[
          'Lottery_Number',
          'Application_Submission_Type',
          'Application_Submitted_Date',
          'Application_Language',
          'Total_Household_Size',
          'Total_Monthly_Rent',
          'Status',
          'Referral_Source',
          'CreatedBy.Name'
        ]}
        labelMapper={fields}
      />
      <ApplicationDetailsContentCard
        dataCollection={application.Applicant}
        title='Primary Applicant'
        fields={[
          'First_Name',
          'Last_Name',
          'Date_of_Birth',
          'Phone',
          'Second_Phone',
          'Email',
          'Residence_Address',
          'Mailing_Address',
        ]}
      />
      <ApplicationDetailsContentCard
        dataCollection={application.Alternate_Contact}
        title='Alternate Contact'
        fields={[
          'First_Name',
          'Last_Name',
          'Phone',
          'Agency_Name',
          'Email',
          'Alternate_Contact_Type',
          'Alternate_Contact_Type_Other'
        ]}
      />
      {(() => {
        if (application.household_members) {
          return (
            <ApplicationDetailsContentTable
              data={application}
              title='Household Members'
              table='household_members'
              fields={[
                'Name',
                'Relationship_to_Applicant',
                'Date_of_Birth',
                'Street',
                'City',
                'State',
                'Zip_Code'
              ]}
            />
          )
        }
      })()}
      <ApplicationDetailsContentCard
        dataCollection={application}
        title='Reserved and Priority Qualifying Information'
        fields={[
          'Has_Military_Service',
          'Has_DevelopmentalDisability',
          'Has_ADA_Priorities_Selected',
          'Answered_Community_Screening'
        ]}
        labelMapper={fields}
      />
      {(() => {
        if (application.preferences) {
          return (
            <ApplicationDetailsContentTable
              data={application}
              title='Application Preferences'
              table='preferences'
              fields={[
                'Receives_Preference',
                'Preference_Name',
                'Lottery_Status',
                'Person_who_claimed_Name',
                'Type_of_proof',
                'Preference_Lottery_Rank',
                'Opt_Out'
              ]}
            />
          )
        }
      })()}
      <ApplicationDetailsContentCard
        dataCollection={application}
        title='Declared Household Income'
        fields={[
          'Annual_Income',
          'Monthly_Income',
          'Housing_Voucher_or_Subsidy'
        ]}
      />
      {(() => {
        if (application.flagged_applications) {
          return (
            <ApplicationDetailsContentTable
              data={application}
              title='Flagged Applications'
              table='flagged_applications'
              fields={[
                'Flagged_Record_Set.Rule_Name',
                'Flagged_Record_Set.Total_Number_of_Pending_Review',
                'View Record Set'
              ]}
            />
          )
        }
      })()}
      <div className="content-card">
        <h4 className="content-card_title t-serif">Attachments</h4>
        <ul>{proofList}</ul>
      </div>
    </div>
  )
}

export default ApplicationDetails
