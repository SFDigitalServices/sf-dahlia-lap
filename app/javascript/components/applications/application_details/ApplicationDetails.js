import React from 'react'

import { map, isEmpty } from 'lodash'

import appPaths from 'utils/appPaths'

import ApplicationDetailsContentCard from './ApplicationDetailsContentCard'
import ApplicationDetailsContentTable from './ApplicationDetailsContentTable'
import {
  flaggedApplicationsFields,
  declareHouseholdIncome,
  applicationDataFields,
  primaryApplicantFields,
  alternateContactFields,
  householdMembersFields,
  applicationPreferencesFields,
  reservedAndPriorityFields,
  applicationEligibilityFields
} from './fields'

const ProofListItems = ({ fileBaseUrl, files }) =>
  map(files, (file) => {
    return (
      <li key={file.id}>
        <a target='_blank' rel='noreferrer' href={appPaths.toAttachmentDownload(fileBaseUrl, file)}>
          {file.document_type}
        </a>
      </li>
    )
  })

const ApplicationDetails = ({ application, fields, fileBaseUrl }) => {
  const ApplicationCard = (props) => (
    <ApplicationDetailsContentCard dataCollection={application} {...props} />
  )
  const Table = (props) => <ApplicationDetailsContentTable data={application} {...props} />

  return (
    <div className='application-details'>
      <ApplicationCard
        title='Application Data'
        fields={applicationDataFields}
        labelMapper={fields}
      />
      <ApplicationDetailsContentCard
        dataCollection={application.applicant}
        title='Primary Applicant'
        fields={primaryApplicantFields}
      />
      {application.listing.is_sale && (
        <ApplicationDetailsContentCard
          title='Eligibility'
          dataCollection={application}
          fields={applicationEligibilityFields}
          splitOn={3}
          labelMapper={fields}
        />
      )}
      <ApplicationDetailsContentCard
        dataCollection={application.alternate_contact}
        title='Alternate Contact'
        fields={alternateContactFields}
      />
      {!isEmpty(application.household_members) && (
        <Table
          title='Household Members'
          table='household_members'
          fields={householdMembersFields}
        />
      )}
      <ApplicationCard
        title='Reserved and Priority Qualifying Information'
        fields={reservedAndPriorityFields}
        labelMapper={fields}
      />
      {!isEmpty(application.preferences) && (
        <Table
          title='Application Preferences'
          table='preferences'
          fields={applicationPreferencesFields}
        />
      )}
      <ApplicationCard title='Declared Household Income' fields={declareHouseholdIncome} />
      {!isEmpty(application.flagged_applications) && (
        <Table
          title='Flagged Applications'
          table='flagged_applications'
          fields={flaggedApplicationsFields}
        />
      )}
      <div className='content-card'>
        <h4 className='content-card_title t-serif'>Attachments</h4>
        <ul>
          <ProofListItems files={application.proof_files} fileBaseUrl={fileBaseUrl} />
        </ul>
      </div>
    </div>
  )
}

export default ApplicationDetails
