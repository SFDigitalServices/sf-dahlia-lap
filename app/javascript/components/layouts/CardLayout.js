import React, { useEffect } from 'react'

import PageHeader from '../organisms/PageHeader'
import TabsSection from '../organisms/TabsSection'
import TabCard from '../organisms/TabCard'
import AppCard from '../molecules/AppCard'

const CardLayout = ({ children, pageHeader, tabSection, toolbar }) => {
  // With the empty array passed as the second argument
  // this useEffect call acts like a componentDidMount call
  useEffect(() => {
    document.body.classList.add('bg-snow')
  }, [])

  return (
    <>
      <PageHeader {...pageHeader} background='snow' />
      { tabSection
        ? (
          <TabsSection {...tabSection} padding='false'>
            {children}
          </TabsSection>
        )
        : (
          <>
            { toolbar && toolbar() }
            <TabCard padding>
              <AppCard>{children}</AppCard>
            </TabCard>
          </>
        )
      }
    </>
  )
}

export default CardLayout
