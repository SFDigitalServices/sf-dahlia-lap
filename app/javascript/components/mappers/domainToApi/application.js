import { toPairs, filter, map, join, compact } from 'lodash'
import { createFieldMapper } from '~/utils/objectUtils'
const checkListToString = (list) => {
  return join(compact(list), ';')
}

const adaPrioritiesMapValues = {
  'vision_impairments': 'Vision impairments',
  'mobility_impairments': 'Mobility impairments',
  'hearing_impairments': 'Hearing impairments'
}

const mapAdaPrioritiesMap = (list) => {
  const selectedPriorities = filter(toPairs(list), ([a, b]) => b)
  return map(selectedPriorities, ([a, b]) => adaPrioritiesMapValues[a])
}

export const applicationFieldMapper = {
  id: 'id',
  primary_applicant_contact: 'primaryApplicantContact',
  application_language: 'applicationLanguage',
  has_military_service: 'hasMilitaryService',
  has_developmental_disability: 'hasDevelopmentalDisability',
  answered_community_screening: 'answeredCommunityScreening',
  housing_voucher_or_subsidy: 'householdVouchersSubsidies',
  terms_acknowledged: 'agreeToTerms',
  number_of_dependents: 'numberOfDependents',
  number_of_seniors: 'numberOfSeniors',
  number_of_minors: 'numberOfMinors',
  reserved_senior: 'hasSenior',
  application_submission_type: 'applicationSubmissionType',
  processing_status: 'processingStatus',
  household_assets: 'householdAssets',
  annual_income: 'annualIncome',
  monthly_income: 'monthlyIncome',
  confirmed_household_annual_income: 'confirmedHouseholdAnnualIncome',
  hh_total_income_with_assets_annual: 'HHTotalIncomeWithAssets',
  total_monthly_rent: 'totalMonthlyRent',
  adaPrioritiesSelected: (source) => checkListToString(mapAdaPrioritiesMap(source.has_ada_priorities_selected)),
  is_first_time_homebuyer: 'isFirstTimeHomebuyer',
  has_completed_homebuyer_education: 'hasCompletedHomebuyerEducation',
  has_loan_preapproval: 'hasLoanPreapproval',
  lending_agent: 'lendingAgent'
}

export const mapApplication = createFieldMapper(applicationFieldMapper)
