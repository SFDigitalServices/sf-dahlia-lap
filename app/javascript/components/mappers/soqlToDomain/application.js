import { split, snakeCase, fromPairs } from 'lodash'
import { mapApplicationMember } from './application_member'
import { mapApplicationPreference } from './application_preference'
import { mapAttachment } from './attachment'
import { mapDemographics } from './demographics'
import { mapFlaggedApplication } from './flagged_application'
import { mapListing } from './listing'
import { mapUser } from './user'
import { mapShape, mapList } from '../utils'
import formUtils from '~/utils/formUtils'

const parseList = text => split(text, ';')
const toChecklist = list => fromPairs(list.map(i => [snakeCase(i), true]))

export const mapApplication = (a) => {
  return {
    applicant: mapShape(mapApplicationMember, a.Applicant),
    demographics: mapShape(mapDemographics, a.Applicant),
    primary_applicant_contact: a.Primary_Applicant,
    alternate_contact: mapShape(mapApplicationMember, a.Alternate_Contact),
    listing: mapShape(mapListing, a.Listing),
    preferences: mapList(mapApplicationPreference, a.preferences),
    proof_files: mapList(mapAttachment, a.proof_files),
    household_members: mapList(mapApplicationMember, a.household_members),
    flagged_applications: mapList(mapFlaggedApplication, a.flagged_applications),
    lease: a.lease,
    rental_assistances: a.rental_assistances,
    createdby: mapShape(mapUser, a.CreatedBy),
    id: a.Id,
    name: a.Name,
    status: a.Status,
    total_household_size: a.Total_Household_Size,
    application_submission_type: a.Application_Submission_Type,
    application_submitted_date: a.Application_Submitted_Date,
    annual_income: formUtils.formatPrice(a.Annual_Income),
    confirmed_household_annual_income: formUtils.formatPrice(a.Confirmed_Household_Annual_Income),
    hh_total_income_with_assets_annual: formUtils.formatPrice(a.HH_Total_Income_with_Assets_Annual),
    household_assets: formUtils.formatPrice(a.Household_Assets),
    monthly_income: a.Monthly_Income,
    housing_voucher_or_subsidy: a.Housing_Voucher_or_Subsidy,
    referral_source: a.Referral_Source,
    application_language: a.Application_Language,
    lottery_number_manual: a.Lottery_Number_Manual,
    lottery_number: a.Lottery_Number,
    total_monthly_rent: formUtils.formatPrice(a.Total_Monthly_Rent),
    general_lottery: a.General_Lottery,
    general_lottery_rank: a.General_Lottery_Rank,
    answered_community_screening: a.Answered_Community_Screening,
    has_military_service: a.Has_Military_Service,
    has_developmental_disability: a.Has_DevelopmentalDisability,
    has_ada_priorities_selected: toChecklist(parseList(a.Has_ADA_Priorities_Selected)),
    terms_acknowledged: a.Terms_Acknowledged,
    number_of_dependents: a.Number_of_Dependents,
    number_of_seniors: a.Number_of_Seniors,
    number_of_minors: a.Number_of_Minors,
    processing_status: a.Processing_Status,
    status_last_updated: a.Status_Last_Updated,
    reserved_senior: a.Reserved_Senior,
    is_first_time_homebuyer: a.Is_First_Time_Homebuyer,
    has_completed_homebuyer_education: a.Has_Completed_Homebuyer_Education,
    has_loan_preapproval: a.Has_Loan_Pre_approval,
    lending_agent: a.Lending_Agent
  }
}
