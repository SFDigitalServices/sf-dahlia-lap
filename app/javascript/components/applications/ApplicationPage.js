import React, { useState } from 'react'
import { isEmpty } from 'lodash'
import { useEffectOnMount } from '~/utils/customHooks'
import { getShortFormApplication } from '~/components/lease_ups/shortFormActions'
import ApplicationDetails from './application_details/ApplicationDetails'
import CardLayout from '../layouts/CardLayout'
import appPaths from '~/utils/appPaths'
import Loading from '~/components/molecules/Loading'

import labelMapperFields from './application_details/applicationDetailsFieldsDesc'

const buildActionLinkIfNecessary = (app, showAddBtn) => {
  const actions = []

  if (!app.listing.is_lottery_complete && app.application_submission_type === 'Paper') {
    actions.push(
      <a
        key='edit-application'
        href={appPaths.toApplicationEdit(app.id)}
        className='primary button tiny'
      >
        Edit Application
      </a>
    )
  }

  if (showAddBtn === 'true') {
    actions.push(
      <a
        key='add-new-application'
        href={appPaths.toApplicationNew(app.listing.id)}
        className='button tiny margin-left--half'
      >
        Add new application
      </a>
    )
  }

  return actions
}

const ApplicationPage = ({ applicationId, showAddBtn, isLeaseUp }) => {
  const [application, setApplication] = useState(null)
  const [fileBaseUrl, setFileBaseUrl] = useState(null)
  const [loading, setLoading] = useState(true)

  useEffectOnMount(() => {
    getShortFormApplication(applicationId)
      .then((response) => {
        setApplication(response.application)
        setFileBaseUrl(response.fileBaseUrl)
      })
      .catch((e) => {
        // Alert window pauses state updates so we set loading to false instead of
        // waiting for the finally block
        setLoading(false)
        window.alert('The application you requested could not be found.')
      })
      .finally(() => setLoading(false))
  })

  let pageHeader = {}
  let tabSection = false

  if (!application) {
    pageHeader = {
      title: 'Application',
      content: <span>Name of Listing:</span>
    }
    // If the lease_up=true param is passed in the url, then we should display the application as if it's
    // a part of the lease up section.
  } else if (isLeaseUp) {
    pageHeader = {
      title: `${application.name}: ${application.applicant.name}`,
      breadcrumbs: [
        { title: 'Lease Ups', link: appPaths.toLeaseUps() },
        {
          title: application.listing.name,
          link: appPaths.toListingLeaseUps(application.listing.id)
        },
        { title: application.name, link: '#' }
      ]
    }

    tabSection = {
      items: [
        {
          title: 'Short Form Application',
          url: appPaths.toLeaseUpShortForm(application.id),
          active: true
        },
        {
          title: 'Supplemental Information',
          url: appPaths.toApplicationSupplementals(application.id)
        }
      ]
    }
  } else {
    pageHeader = {
      title: `Application ${application.name}`,
      content: (
        <span>
          Name of Listing:{' '}
          <a href={appPaths.toListing(application.listing.id)}>{application.listing.name}</a>
        </span>
      ),
      action: buildActionLinkIfNecessary(application, showAddBtn)
    }
  }

  return (
    <Loading isLoading={loading}>
      <CardLayout pageHeader={pageHeader} tabSection={tabSection}>
        {!isEmpty(application) && (
          <ApplicationDetails
            application={application}
            fileBaseUrl={fileBaseUrl}
            fields={labelMapperFields}
          />
        )}
      </CardLayout>
    </Loading>
  )
}

export default ApplicationPage
