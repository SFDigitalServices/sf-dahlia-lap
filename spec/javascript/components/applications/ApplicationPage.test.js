import { shallow } from 'enzyme'
import { act } from 'react-dom/test-utils'

import ApplicationDetails from 'components/applications/application_details/ApplicationDetails'
import ApplicationPage from 'components/applications/ApplicationPage'
import CardLayout from 'components/layouts/CardLayout'
import Loading from 'components/molecules/Loading'

import application from '../../fixtures/application'
import saleApplication from '../../fixtures/sale_application'
import { mountAppWithUrl } from '../../testUtils/wrapperUtil'

const flaggedAppCardSelector = '#content-card-flagged_applications'

const FLAGGED_APP_ID = 'flagged_app_id'
const NO_FLAGS_ID = 'no_flags_id'
const SALE_APP_ID = 'sale_app_id'
const REGULAR_APP_ID = 'regular_app_id'

const mockGetShortFormApplication = jest.fn()

const tick = () => {
  return new Promise((resolve) => {
    setTimeout(resolve, 0)
  })
}

jest.mock('apiService', () => {
  const FLAGGED_APP_ID = 'flagged_app_id'
  const NO_FLAGS_ID = 'no_flags_id'
  const SALE_APP_ID = 'sale_app_id'
  return {
    getShortFormApplication: async (id) => {
      mockGetShortFormApplication(id)
      let responseApp
      if (id === FLAGGED_APP_ID) {
        responseApp = {
          ...mockApplication,
          flagged_applications: [
            {
              flagged_record: {
                id: 'a0r0P00002X4r08QAB',
                rule_name: 'Name + DOB',
                total_number_of_pending_review: 51
              }
            }
          ]
        }
      } else if (id === SALE_APP_ID) {
        responseApp = { ...mockSaleApplication }
      } else if (id === NO_FLAGS_ID) {
        responseApp = { ...mockApplication, flagged_applications: [] }
      } else {
        responseApp = { ...mockApplication }
      }

      return Promise.resolve({ application: responseApp, fileBaseUrl: 'test_file_url' })
    }
  }
})

// we need to define these separate from the import so we can use them in the
// jest block.
const mockApplication = application
const mockSaleApplication = saleApplication

const getWrapper = async ({
  applicationId,
  withLeaseUpUrl = false,
  waitForLoadingFinish = false,
  queryParams = ''
}) => {
  const url = withLeaseUpUrl
    ? `/lease-ups/applications/${applicationId}${queryParams}`
    : `/applications/${applicationId}${queryParams}`

  let wrapper

  // have to wrap in await act so that all useEffectOnMount updates finish.
  await act(async () => {
    wrapper = mountAppWithUrl(url)
  })

  if (waitForLoadingFinish) {
    await act(tick)
    wrapper.update()
  }

  return wrapper.find(ApplicationPage)
}

describe('ApplicationPage', () => {
  describe('with regular application', () => {
    const appId = REGULAR_APP_ID
    describe('before application is loaded', () => {
      let wrapper
      beforeEach(async () => {
        wrapper = await getWrapper({ applicationId: appId })
      })

      test('should be loading', () => {
        expect(wrapper.find(Loading).prop('isLoading')).toBeTruthy()
      })

      test('renders the correct header title and content', () => {
        expect(wrapper.find(CardLayout).props().pageHeader.title).toEqual('Application ')

        const headerContentWrapper = shallow(wrapper.find(CardLayout).props().pageHeader.content)
        expect(headerContentWrapper.text()).toEqual('Name of Listing: ')
      })

      test('does not render the tab section', () => {
        expect(wrapper.find(CardLayout).props().tabSection).toBeUndefined()
      })

      test('does not render ApplicationDetails', () => {
        expect(wrapper.find(ApplicationDetails)).toHaveLength(0)
      })

      test('called getShortFormApplication', () => {
        expect(mockGetShortFormApplication.mock.calls).toHaveLength(1)
        expect(mockGetShortFormApplication.mock.calls[0][0]).toEqual(REGULAR_APP_ID)
      })
    })

    describe('After the application finishes loading', () => {
      let wrapper
      beforeEach(async () => {
        wrapper = await getWrapper({ applicationId: appId, waitForLoadingFinish: true })
      })

      test('should no longer be loading', () => {
        expect(wrapper.find(Loading).prop('isLoading')).toBeFalsy()
      })

      test('renders ApplicationDetails', () => {
        expect(wrapper.find(ApplicationDetails)).toHaveLength(1)
      })

      test('populates the header title and content with application details', () => {
        expect(wrapper.find(CardLayout).props().pageHeader.title).toEqual(
          'Application APP-00191270'
        )

        const headerContentWrapper = shallow(wrapper.find(CardLayout).props().pageHeader.content)
        expect(headerContentWrapper.text()).toEqual('Name of Listing: Test 5/30')
      })

      test('does not render the tab section', () => {
        expect(wrapper.find(CardLayout).props().tabSection).toBeUndefined()
      })

      test('should match snapshot', () => {
        // TODO: Expand test coverage on this page to the point that we do not need this snapshot check.
        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('with sale application', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({ applicationId: SALE_APP_ID, waitForLoadingFinish: true })
    })

    test('renders Eligibility section', async () => {
      expect(wrapper.find('.content-card_title').at(2).text()).toEqual('Eligibility')
    })

    test('renders name of lender', async () => {
      expect(wrapper.find('.application-details').childAt(2).text()).toContain('Jason Lockhart')
    })
  })

  describe('with application that has no flags', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({ applicationId: NO_FLAGS_ID, waitForLoadingFinish: true })
    })

    test('does not render flagged app content card', async () => {
      expect(wrapper.exists(flaggedAppCardSelector)).toEqual(false)
    })
  })

  describe('with flagged application', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({ applicationId: FLAGGED_APP_ID, waitForLoadingFinish: true })
    })

    test('renders flagged app content card', async () => {
      expect(wrapper.exists(flaggedAppCardSelector)).toEqual(true)
    })

    test('renders the correct rule name', () => {
      expect(wrapper.find(`${flaggedAppCardSelector} > table > tbody > tr`)).toHaveLength(1)
      expect(
        wrapper.find(`${flaggedAppCardSelector} > table > tbody > tr > td`).first().text()
      ).toEqual('Name + DOB')
    })
  })

  describe('with all query params = null', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({
        applicationId: REGULAR_APP_ID,
        waitForLoadingFinish: true,
        queryParams: ''
      })
    })

    test('does not render the tab headers', async () => {
      expect(wrapper.text()).not.toContain('Supplemental Information')
      expect(wrapper.text()).not.toContain('Short Form Application')
    })

    test('does not render add button link', async () => {
      expect(wrapper.text()).not.toContain('Add new application')
    })
  })

  describe('without lease-up URL', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({
        applicationId: REGULAR_APP_ID,
        waitForLoadingFinish: true
      })
    })

    test('does not render the tab headers', async () => {
      expect(wrapper.text()).not.toContain('Supplemental Information')
      expect(wrapper.text()).not.toContain('Short Form Application')
    })
  })

  describe('with leaseUp url', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({
        applicationId: REGULAR_APP_ID,
        withLeaseUpUrl: true,
        waitForLoadingFinish: true
      })
    })

    test('renders the tab headers', async () => {
      expect(wrapper.text()).toContain('Supplemental Information')
      expect(wrapper.text()).toContain('Short Form Application')
    })
  })

  describe('with showAppBtn query param = false', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({
        applicationId: REGULAR_APP_ID,
        waitForLoadingFinish: true,
        queryParams: '?showAddBtn=false'
      })
    })

    test('does not render add button link', async () => {
      expect(wrapper.text()).not.toContain('Add new application')
    })
  })

  describe('with showAppBtn query param = true', () => {
    let wrapper
    beforeEach(async () => {
      wrapper = await getWrapper({
        applicationId: REGULAR_APP_ID,
        waitForLoadingFinish: true,
        queryParams: '?showAddBtn=true'
      })
    })

    test('renders add button link', async () => {
      expect(wrapper.text()).toContain('Add new application')
    })
  })

  describe('with leaseup url and showAppBtn', () => {
    describe('before app is loaded', () => {
      let wrapper
      beforeEach(async () => {
        wrapper = await getWrapper({
          applicationId: REGULAR_APP_ID,
          withLeaseUpUrl: true,
          waitForLoadingFinish: false,
          queryParams: '?showAddBtn=true'
        })
      })

      test('renders the correct header title and content', () => {
        const titleText = shallow(wrapper.find(CardLayout).props().pageHeader.title)
        const nbspUnicode = '\u00a0'
        expect(titleText.text()).toEqual(nbspUnicode)

        expect(
          wrapper
            .find(CardLayout)
            .props()
            .pageHeader.breadcrumbs.map((crumb) => crumb.title)
        ).toEqual(['Lease Ups', '', ''])
      })

      test('renders the tab section', () => {
        expect(
          wrapper
            .find(CardLayout)
            .props()
            .tabSection.items.map((i) => i.title)
        ).toEqual(['Short Form Application', 'Supplemental Information'])
      })

      test('does not render add button link', async () => {
        // add button link behavior isn't possible with lease up view
        expect(wrapper.text()).not.toContain('Add new application')
      })

      test('renders the tab headers', async () => {
        expect(wrapper.text()).toContain('Supplemental Information')
        expect(wrapper.text()).toContain('Short Form Application')
      })
    })

    describe('after loading finish', () => {
      let wrapper
      beforeEach(async () => {
        wrapper = await getWrapper({
          applicationId: REGULAR_APP_ID,
          withLeaseUpUrl: true,
          waitForLoadingFinish: true,
          queryParams: '?showAddBtn=true'
        })
      })

      test('renders the correct header title and content', () => {
        expect(wrapper.find(CardLayout).props().pageHeader.title).toEqual(
          'APP-00191270: karen elizabeth jones'
        )

        expect(
          wrapper
            .find(CardLayout)
            .props()
            .pageHeader.breadcrumbs.map((crumb) => crumb.title)
        ).toEqual(['Lease Ups', 'Test 5/30', 'APP-00191270'])
      })

      test('renders the tab section', () => {
        expect(
          wrapper
            .find(CardLayout)
            .props()
            .tabSection.items.map((i) => i.title)
        ).toEqual(['Short Form Application', 'Supplemental Information'])
      })

      test('does not render add button link', async () => {
        // add button link behavior isn't possible with lease up view
        expect(wrapper.text()).not.toContain('Add new application')
      })

      test('renders the tab headers', async () => {
        expect(wrapper.text()).toContain('Supplemental Information')
        expect(wrapper.text()).toContain('Short Form Application')
      })
    })
  })
})
