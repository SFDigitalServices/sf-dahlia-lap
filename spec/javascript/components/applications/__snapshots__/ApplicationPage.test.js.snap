// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationPage should render succesfully 1`] = `
Array [
  <header
    className="lead-header short  bg-snow"
  >
    <div
      className="row full-width inner--3x"
    >
      <div
        className="large-12 columns"
      >
        <hgroup
          className="lead-header_group"
        >
          <h1
            className="lead-header_title small-serif c-oil"
          >
            Application undefined
          </h1>
          <div
            className="medium-8 columns no-padding"
          >
            <p>
              <span>
                Name of Listing: 
                <a
                  href="/listings/1"
                />
              </span>
            </p>
          </div>
        </hgroup>
      </div>
    </div>
  </header>,
  <section
    className="tabs-card-row row full-width inner--3x margin-bottom--2x padding-top--2x"
  >
    <div
      className="app-card form-card tabs-card max-width expand-on-small"
    >
      <div>
        <div
          className="content-card padding-bottom-none margin-bottom--half bg-trans"
        >
          <h4
            className="content-card_title t-serif"
          >
            Application Data
          </h4>
          <ul
            className="content-grid"
          >
            <li
              className="content-item"
            >
              <div
                className="content-card"
              >
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Lottery Number
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Application Submission Type
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Application Submitted Date
                  </h4>
                  <p>
                    04/23/2018
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Application Language
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Total Household Size
                  </h4>
                  <p>
                    None
                  </p>
                </div>
              </div>
            </li>
            <li
              className="content-item"
            >
              <div
                className="content-card"
              >
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Total Monthly Rent
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Status
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Referral Source
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    CreatedBy
                  </h4>
                  <p>
                    None
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
        <div
          className="content-card padding-bottom-none margin-bottom--half bg-trans"
        >
          <h4
            className="content-card_title t-serif"
          >
            Primary Applicant
          </h4>
          <ul
            className="content-grid"
          >
            <li
              className="content-item"
            >
              <div
                className="content-card"
              />
            </li>
            <li
              className="content-item"
            >
              <div
                className="content-card"
              />
            </li>
          </ul>
        </div>
        <div
          className="content-card padding-bottom-none margin-bottom--half bg-trans"
        >
          <h4
            className="content-card_title t-serif"
          >
            Alternate Contact
          </h4>
          <ul
            className="content-grid"
          >
            <li
              className="content-item"
            >
              <div
                className="content-card"
              />
            </li>
            <li
              className="content-item"
            >
              <div
                className="content-card"
              />
            </li>
          </ul>
        </div>
        <div
          className="content-card padding-bottom-none margin-bottom--half bg-trans"
        >
          <h4
            className="content-card_title t-serif"
          >
            Reserved and Priority Qualifying Information
          </h4>
          <ul
            className="content-grid"
          >
            <li
              className="content-item"
            >
              <div
                className="content-card"
              >
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Has Military Service
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Has DevelopmentalDisability
                  </h4>
                  <p>
                    None
                  </p>
                </div>
              </div>
            </li>
            <li
              className="content-item"
            >
              <div
                className="content-card"
              >
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Has ADA Priorities Selected
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Answered Community Screening
                  </h4>
                  <p>
                    None
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
        <div
          className="content-card padding-bottom-none margin-bottom--half bg-trans"
        >
          <h4
            className="content-card_title t-serif"
          >
            Declared Household Income
          </h4>
          <ul
            className="content-grid"
          >
            <li
              className="content-item"
            >
              <div
                className="content-card"
              >
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Annual Income
                  </h4>
                  <p>
                    None
                  </p>
                </div>
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Monthly Income
                  </h4>
                  <p>
                    None
                  </p>
                </div>
              </div>
            </li>
            <li
              className="content-item"
            >
              <div
                className="content-card"
              >
                <div
                  className="margin-bottom--half"
                >
                  <h4
                    className="t-sans t-small t-bold no-margin"
                  >
                    Housing Voucher or Subsidy
                  </h4>
                  <p>
                    None
                  </p>
                </div>
              </div>
            </li>
          </ul>
        </div>
        <div
          className="content-card"
        >
          <h4
            className="content-card_title t-serif"
          >
            Attachments
          </h4>
          <ul />
        </div>
      </div>
    </div>
  </section>,
]
`;
